contrib/remotexact/apply.c

```diff
@@ -0,0 +1,369 @@
+static void apply_handle_update(StringInfo s, bool skip);
+static void apply_handle_delete(StringInfo s, bool skip);
+static Relation open_relation(LogicalRepRelId relid, LOCKMODE lockmode);
+static void close_relation(Relation rel, LOCKMODE lockmode);
+static ApplyExecutionData *create_edata_for_relation(Relation rel);
+static void finish_edata(ApplyExecutionData *edata);
+static void slot_store_data(TupleTableSlot *slot,
+                            Relation rel,
+				            LogicalRepTupleData *tupleData);
+
+void
+apply_writes(RWSet *rwset)
+{
+	StringInfoData s;
+
+	s.data = rwset->writes;
+	s.len = rwset->writes_len;
+	s.cursor = 0;
+	while (s.cursor < s.len)
+	{
+		int region;
+		LogicalRepMsgType action;
+		bool skip;
+
+		region = pq_getmsgbyte(&s);
+
+		// We ignore tuples that do not belong to the current region
+		// by setting the skip argument to true to the apply functions
+		// below. They will still decode the tuples but will not apply
+		// the changes for them.
+		skip = region != current_region;
+
+		action = pq_getmsgbyte(&s);
+		switch (action)
+		{
+			case LOGICAL_REP_MSG_INSERT:
+				apply_handle_insert(&s, skip);
+				break;
+
+			case LOGICAL_REP_MSG_UPDATE:
+				apply_handle_update(&s, skip);
+				break;
+
+			case LOGICAL_REP_MSG_DELETE:
+				apply_handle_delete(&s, skip);
+				break;
+
+			default:
+				ereport(ERROR,
+						(errcode(ERRCODE_PROTOCOL_VIOLATION),
+						 errmsg_internal("[remotexact] invalid write set message type \"%c\"", action)));
+		}
+	}
+}
+
+/*
+ * Ported from apply_handle_insert in backend/replication/logical/worker.c
+ */
+static void
+apply_handle_insert(StringInfo s, bool skip)
+{
+	Relation rel;
+	LogicalRepTupleData newtup;
+	LogicalRepRelId relid;
+	ApplyExecutionData *edata;
+	EState	   *estate;
+	TupleTableSlot *remoteslot;
+	MemoryContext oldctx;
+
+	relid = logicalrep_read_insert(s, &newtup);
+
+	if (skip)
+		return;
+
+	rel = open_relation(relid, RowExclusiveLock);
+
+	/* Initialize the executor state. */
+	edata = create_edata_for_relation(rel);
+	estate = edata->estate;
+	remoteslot = ExecInitExtraTupleSlot(estate,
+										RelationGetDescr(rel),
+										&TTSOpsVirtual);
+
+	/* Process and store remote tuple in the slot */
+	oldctx = MemoryContextSwitchTo(GetPerTupleMemoryContext(estate));
+	slot_store_data(remoteslot, rel, &newtup);
+	MemoryContextSwitchTo(oldctx);
+
+	/* Unlike logical replication, we never write to a partitioned relation
+	   (but the relation can be  a partition of a partitioned relation), so
+	   we don't need to check whether the relation is partitioned or not. */
+
+	/* We must open indexes here. */
+	ExecOpenIndices(edata->targetRelInfo, false);
+
+	/* Do the insert. */
+	ExecSimpleRelationInsert(edata->targetRelInfo, edata->estate, remoteslot);
+
+	/* Cleanup. */
+	ExecCloseIndices(edata->targetRelInfo);
+
+	finish_edata(edata);
+
+	close_relation(rel, NoLock);
+}
+
+static void
+apply_handle_update(StringInfo s, bool skip)
+{
```

Comment:
nit: Add a TODO?

Reply:
I have already implemented this function. PR coming soon.